crates/sargon/src/profile/mfa/security_structures/automatic_shield_builder/automatic_shield_builder.rs

```diff
@@ -0,0 +1,916 @@
+            }
+        }
+    }
+
+    /// Calls `assign_factor_of_category_to_role_while_meaningful_and_less_than_limit`
+    /// for both Custodian and Hardware categories.
+    fn assign_custodian_and_hardware_to_role_while_meaningful_and_less_than_limit(
+        &mut self,
+        limit: u8,
+        to: RoleKind,
+    ) {
+        self.assign_factor_of_category_to_role_while_meaningful_and_less_than_limit(
+            Custodian,
+            limit,
+            to,
+        );
+
+        self.assign_factor_of_category_to_role_while_meaningful_and_less_than_limit(
+            Hardware,
+            limit,
+            to,
+        );
+    }
+
+    /// Calls `assign_custodian_and_hardware_to_role_while_meaningful_and_less_than_limit`
+    /// for both Recovery and Confirmation roles.
+    fn assign_custodian_and_hardware_to_non_primary_roles_while_less_than_limit_for_each_assignment(
+        &mut self,
+        limit: u8,
+    ) {
+        self.assign_custodian_and_hardware_to_role_while_meaningful_and_less_than_limit(
+            limit, Recovery,
+        );
+        self.assign_custodian_and_hardware_to_role_while_meaningful_and_less_than_limit(
+            limit,
+            Confirmation,
+        );
+    }
+
+    /// Automatic assignment of factors to roles according to [this heuristics][doc].
+    ///
+    /// [doc]: https://radixdlt.atlassian.net/wiki/spaces/AT/pages/3758063620/MFA+Rules+for+Factors+and+Security+Shields#Automatic-Security-Shield-Construction
+    fn assign(&mut self) -> Result<ProtoMatrix> {
+        // 📒 "If the user only chose 1 factor for PRIMARY, remove that factor from the list (it cannot be used elsewhere - otherwise it can)."
+        {
+            if self.count_factors_for_role(Primary) == 1
+                && let Some(only_primary_factor) =
+                    self.proto_matrix.primary.iter().next()
+            {
+                self.remaining_available_factors
+                    .retain(|f| f.id() != *only_primary_factor);
+            }
+        }
+
+        // 📒 "Drop in the somewhat “special-use” factors first"
+        {
+            // 📒 "Add all Contact factors in the list to RECOVERY."
+            self.assign_factors_of_category_to_recovery(Contact, Quantity::All);
+
+            // 📒	"Add all Information factors in the list to CONFIRMATION."
+            self.assign_factors_of_category_to_confirmation(
+                Information,
+                Quantity::All,
+            );
+        }
+
+        // 📒 Assign Custodian/Hardware factors to RECOVERY & CONFIRMATION
+        // without exceeding limit of 1 factor in each role.
+        self.assign_custodian_and_hardware_to_non_primary_roles_while_less_than_limit_for_each_assignment(1);
+
+        // 📒 Assign Custodian/Hardware factors to RECOVERY & CONFIRMATION
+        // without exceeding limit of 2 factor in each role.
+        self.assign_custodian_and_hardware_to_non_primary_roles_while_less_than_limit_for_each_assignment(2);
```

Comment:
A bit hard to understand why the second call will not cover the first one also, can you elaborate?

Reply:
let me know if you have concrete suggestions, now that we have walked this through in a call.